python/shark_turbine/runtime/op_reg/base.py

```diff
@@ -0,0 +1,505 @@
+# Copyright 2023 Advanced Micro Devices, Inc
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+"""Base classes for registering custom operations with the PyTorch
+dispatcher.
+"""
+
+from typing import Any, Callable, Optional, Type, Union
+
+from abc import ABC, abstractmethod, abstractproperty
+import functools
+import logging
+
+import torch
+from torch import Tensor
+
+from ...aot.support.ir_imports import (
+    Block,
+    Context,
+    FunctionType,
+    InsertionPoint,
+    Location,
+    Operation,
+    StringAttr,
+    SymbolTable,
+    IrType,
+    Value,
+    builtin_d,
+    func_d,
+)
+
+from ...aot.support.ir_utils import (
+    TORCH_DTYPE_TO_IREE_TYPE_ASM,
+)
+
+__all__ = [
+    "ArgDescriptor",
+    "CustomOp",
+    "FreeFuncKernelBuilder",
+    "IntArg",
+    "KernelBuilder",
+    "KernelGenerator",
+    "KernelSelection",
+    "TensorArg",
+]
+
+logger = logging.getLogger("turbine.runtime.op_reg")
+
+###############################################################################
+# Op library management
+###############################################################################
+
+# All such custom kernels are registered in the 'turbine' library/namespace.
+# We also allow extending existing libraries outside of this, but that is
+# the non default case.
+TURBINE_LIBRARY = torch.library.Library("turbine", "DEF")
+
+
+class CustomOp(ABC):
+    """Users subclass this in order to register a turbine custom op."""
+
+    @staticmethod
+    def register(
+        op_class: Optional[Type["CustomOp"]],
+        *,
+        library: torch.library.Library = TURBINE_LIBRARY,
+        dispatch_key: str = "",
+        register_meta: bool = True,
+        register_impl: bool = True,
+    ) -> Callable:
+        """Class decorator for `CustomOp` implementations.
+
+        The decorator will instantiate the class and then replace it with
+        the callable operation that can be used to invoke the kernel.
+
+        Typical usage:
+
+        ```
+        @CustomOp.register
```

Comment:
I didn't realize you could decorate classes. This must be the way pytorch layers are implemented 🤯

Reply:
One of the many magical things, yes. It was horribly broken in Python 2 so was seldom used until the 3.x transition completed.